eth_utils/abi.py

```diff
@@ -1,68 +1,793 @@
+        new_abi = copy.copy(arg_abi)
+        new_abi["type"] = tuple_prefix + "[]" * (num_dims - 1)
+
+        sub_abis = itertools.repeat(new_abi)
+
+    if isinstance(arg, abc.Mapping):
+        # Arg is mapping.  Align values according to abi order.
+        aligned_arg = tuple(arg[abi["name"]] for abi in sub_abis)
+    else:
+        aligned_arg = arg
+
+    if not is_list_like(aligned_arg):
         raise TypeError(
-            f"The 'type' must be a string, but got {repr(typ)} of type {type(typ)}"
+            f'Expected non-string sequence for "{arg_abi.get("type")}" '
+            f"component type: got {aligned_arg}"
         )
-    elif not typ.startswith("tuple"):
-        return typ
 
-    delimited = ",".join(collapse_if_tuple(c) for c in abi["components"])
-    # Whatever comes after "tuple" is the array dims. The ABI spec states that
-    # this will have the form "", "[]", or "[k]".
-    array_dim = typ[5:]
-    collapsed = f"({delimited}){array_dim}"
+    # convert NamedTuple to regular tuple
+    typing = tuple if isinstance(aligned_arg, tuple) else type(aligned_arg)
 
-    return collapsed
+    return typing(
+        _align_abi_input(sub_abi, sub_arg)
+        for sub_abi, sub_arg in zip(sub_abis, aligned_arg)
+    )
+
+
+def _get_tuple_type_str_and_dims(s: str) -> Optional[Tuple[str, Optional[str]]]:
+    """
+    Takes a JSON ABI type string.  For tuple type strings, returns the separated
+    prefix and array dimension parts.  For all other strings, returns ``None``.
+    """
+    tuple_type_str_re = "^(tuple)((\\[([1-9]\\d*\b)?])*)??$"
+    match = re.compile(tuple_type_str_re).match(s)
+
+    if match is not None:
+        tuple_prefix = match.group(1)
+        tuple_dims = match.group(2)
+
+        return tuple_prefix, tuple_dims
+
+    return None
+
+
+def abi_to_signature(abi_element: ABIElement) -> str:
+    """
+    Returns a string signature representation of the function or event ABI
+    and arguments.
+
+    Signatures consist of the name followed by a list of arguments.
+
+    :param abi_element: ABI element.
+    :type abi_element: `ABIElement`
+    :return: Stringified ABI signature
+    :rtype: `str`
+
+    .. doctest::
+
+        >>> from eth_utils import abi_to_signature
+        >>> abi_element = {
+        ...   'constant': False,
+        ...   'inputs': [
+        ...     {
+        ...       'name': 's',
+        ...       'type': 'uint256'
+        ...     }
+        ...   ],
+        ...   'name': 'f',
+        ...   'outputs': [],
+        ...   'payable': False,
+        ...   'stateMutability': 'nonpayable',
+        ...   'type': 'function'
+        ... }
+        >>> abi_to_signature(abi_element)
+        'f(uint256)'
+    """
+    if abi_element["type"] == "fallback" or abi_element["type"] == "receive":
+        name = str(abi_element["type"])
+        abi_element_inputs = None
+    elif abi_element["type"] == "constructor":
+        name = abi_element["type"]
```

Comment:
Is there a reason this line doesn't use `str()` but does in the clause above?

Reply:
Refactored this logic, I think it should be more explicit now.